plugins/primus_x86/primus_x86_loader.ml

```diff
@@ -50,6 +42,65 @@ module Plt_jumps(Machine : Primus.Machine.S) = struct
 
 end
 
+module Link_plt(Machine : Primus.Machine.S) = struct
+  open Machine.Syntax
+
+  let section_memory sec_name =
+    Machine.get () >>| fun proj ->
+    Memmap.filter (Project.memory proj) ~f:(is_section sec_name) |>
+    Memmap.to_sequence |>
+    Seq.map ~f:fst
+
+  let is_a_plt_stub mem sub =
+    match Term.get_attr sub address with
+    | None -> false
+    | Some addr -> Seq.exists mem ~f:(fun m -> Memory.contains m addr)
+
+  let relink plt prog =
+    let normalize_name x =
+      match String.split ~on:'@' x with
+      | name :: _ -> name
+      | _ -> x in
+    let add m sub =
+      Map.set m (normalize_name @@ Sub.name sub) (Term.tid sub) in
+    let subs = Term.to_sequence sub_t prog in
+    let stubs, subs =
+      Seq.fold ~init:(Map.empty (module String), Map.empty (module String))  subs
+        ~f:(fun (stubs, subs) sub ->
+          if is_a_plt_stub plt sub
+          then add stubs sub, subs
+          else stubs, add subs sub) in
+    let twins =
+      Map.fold stubs ~init:(Map.empty (module Tid))
+        ~f:(fun ~key:name ~data:tid twins ->
+          match Map.find subs name with
+          | None -> twins
+          | Some tid' -> Map.set twins tid tid') in
+    (object
+       inherit Term.mapper
+
+       method! map_jmp jmp =
+         match Jmp.kind jmp with
+         | Goto _ | Int _ | Ret _ -> jmp
+         | Call c ->
+            match Call.target c with
+            | Indirect _ -> jmp
+            | Direct tid ->
+               match Map.find twins tid with
+               | None -> jmp
+               | Some tid' ->
+                  Jmp.with_kind jmp (Call (Call.with_target c (Direct tid')))
+    end)#run prog
```

Comment:
This function is too big, split it into several functions, each having a readable name and well-defined semantics. 

Next, the overall idea of the function is flawed. For your code to work the following assumptions should hold. 
1. all functions have different names (this is enforced by the Program module abstraction, so it is fine);
2. there is a demangler that uses `@` to mangle names that were originally having the same name (it is a hidden assumption that is implementaion defined, we can't rely on it);
3. The original names never had `@` in it. (Doesn't hold at all);
4. A function can have at most one other function that originally had the same name, aka twin in your parlance. (Not true - it could be a set of twins)
5. If a function has a twin then one of this pair must be a stub and another must not (doesn't hold).

So, please, revisit your implementation without relying on these assumptions. My suggestion would be to go through all stubs and try to build a partial mapping between stubs and non-stubs, like your twins mapping. This mapping shall be purely injective. In other words, for each stub there should be at most one implementation, and for each implementation there should be at most one stub. That means that if for a given stub you can select more than one implementation then neither this stub nor this implementation should be taken into the final mapping. The same is for the other side -- if implementation could be selected for more than one stub, then they should also be disregarded. Putting it simple, given a bipartile graph of stubs and implementations, you shall take only those stubs that have _exactly_ one edge. 





[1]: https://en.wikipedia.org/wiki/Partition_of_a_set